src/system/random/basic.lean

```diff
@@ -0,0 +1,524 @@
+import algebra.group_power
+import category.liftable
+
+import data.bitvec
+import data.list.basic
+import data.stream
+import tactic.cache
+-- import util.data.list
+import data.fin
+-- import util.meta.tactic
+
+import tactic.interactive
+import tactic.norm_num
+
+import system.io
+import system.random
+
+open list io applicative
+
+universes u v w
+
+/-- A monad to generate random objects using the generator type `g` -/
+@[reducible]
+def rand_g (g : Type) := state (ulift g)
+/-- A monad to generate random objects using the generator type `std_gen` -/
+@[reducible]
+def rand := rand_g std_gen
+
+instance (g : Type) : liftable (rand_g.{u} g) (rand_g.{v} g) :=
+@state_t.liftable' _ _ _ _ _ _ _ (equiv.ulift.trans.{u u u u u} equiv.ulift.symm)
+
+open ulift
+
+/-- Generate one more `ℕ` -/
+def random.next {gen : Type} [random_gen gen] : rand_g gen ℕ :=
+⟨ prod.map id up ∘ random_gen.next ∘ down ⟩
+
+/-- `range i j` is a number between `i` and `j` inclusively -/
+def range {α : Type u} [has_le α] (i j : α) :=
+{ x : α // i ≤ x ∧ x ≤ j }
+
+infix ` .. `:41 := range
+
+open stream
+
+/-- `random α` gives us machinery to generate values of type `α` -/
+class random (α : Type u) extends has_le α :=
+(random : Π (g : Type) [random_gen g], rand_g g α)
+(random_r : Π g [random_gen g] (x y : α),
+              x ≤ y →
+              rand_g g (x .. y))
+(random_series : Π (gen : Type) [random_gen gen], gen → stream α :=
+by { intros, resetI,
+     exact corec prod.fst ((random gen).run ∘ prod.snd) ( (random gen).run ⟨ a ⟩ ) } )
+(random_series_r : Π (g : Type) [random_gen g] (x y : α)
+                        (h : x ≤ y),
+                        g →
+                        stream (x .. y) :=
+by { introsI,
+     exact corec prod.fst ((random_r g x y h).run ∘ prod.snd) ((random_r g x y h).run ⟨ a ⟩) } )
+
+namespace tactic.interactive
+
+/-- Some functions require a non-empty range as a parameter. This
+tactic checks that the range is non-empty -/
+meta def check_range : tactic unit :=
+assumption <|> do
+`[apply of_as_true, trivial]
+
+end tactic.interactive
+
+export tactic.interactive (check_range)
+
+namespace io
+
+/-- on a platform with a `/dev/random` file to get random bytes from,
+get `n` bytes from it -/
+def read_dev_random (n : ℕ) : io (array n char) := do
+fh ← mk_file_handle "/dev/random" mode.read tt,
```

Comment:
Succeeds on Mac. It will be a problem on Windows. Maybe I should add a primitive in Lean itself to seed a random number generator. Do you see another alternative?

Reply:
I'm not too familiar with QuickCheck or others, but is this something where you'd want a quasirandom number generator anyway instead of actual random numbers?

Reply:
Yes, that's right but I only use this as a way to seed the pseudo-random generator

Reply:
It turns out that the way `io.rand` is implemented, if we don't create our own `std_gen`, it will use the C standard `rand()` function. That's good enough for a seed.